### PR TITLE
feat: PING-3022 setup fetcher and integrate to backend

### PR DESCRIPTION
**Pull request by yusufputra**
https://github.com/BetterSocial/better-webapp-nextjs/assets/13555051/84c34fe4-cca4-408c-84f8-2a8c3ce7826d

todo next PR:
- [x] loading indicator when fetching
- [x] show toast when failed fetching
- [x] add download link
- [ ] followingup link (share, follow)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,13 +34,15 @@
     "react-device-detect": "^2.2.2",
     "react-dom": "^18.2.0",
     "react-helmet": "^6.1.0",
+    "react-query": "^3.39.3",
```
Comment (asaadam) [inline package.json:37]: ada baiknya kita pake `@tanstack/react-query` aja mas ucup

File: services/profile/profileServices.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import api from "shared/fetcher";
+
+const getPublicProfile = async (username: string) => {
```
Comment (asaadam) [inline services/profile/profileServices.ts:3]: ini gak dibungkus react query aja? atau belom selesai? kwkwkw

File: shared/fetcher.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+api.interceptors.response.use(
+  (response) => response,
+  async (error) => {
+    if (
+      error?.response?.status === 401 &&
+      error?.response?.config?.url !== "/users/refresh-token"
+    ) {
+      const token = Cookies.get(ITokenEnum.refreshToken);
+      const refreshApi = axios.create({
```
Comment (asaadam) [inline shared/fetcher.ts:35]: `axios.create` apa ga diganti sama `api.blablabla` aja mas ucup?

File: services/profile/profileServices.ts (hunk 2 above)
Comment (yusufputra) [reply services/profile/profileServices.ts:19]: lho iya tadi perasaan aku pake tanstack hiks, tak ganti sek

File: shared/fetcher.ts (hunk 3 above)
Comment (yusufputra) [reply shared/fetcher.ts:35]: oiya juga masuk bolehh

File: package.json (hunk 1 above)
Comment (yusufputra) [reply package.json:37]: lho iya tadi perasaan aku pake tanstack hiks, tak ganti sek

File: next.config.js [hunk 4]
```diff
@@ -5,6 +5,13 @@ module.exports = {
     domains: [
       'play.google.com',
       'res.cloudinary.com'],
+  },
+  publicRuntimeConfig: {
+    BASE_URL: process.env.BASE_URL || 'https://devme.bettersocial.org',
+    HUMAN_ID_GET_WEB_LOGIN_URL: process.env.HUMAN_ID_GET_WEB_LOGIN_URL || 'https://core.human-id.org/v0.0.3/server/users/web-login',
+    HUMAN_ID_CLIENT_ID: process.env.HUMAN_ID_CLIENT_ID || 'SERVER_83XB8BLZ8GOOTA51H9OFQ5',
+    HUMAN_ID_CLIENT_SECRET: process.env.HUMAN_ID_CLIENT_SECRET || 'ieGo4-YqMa53OGBXOrg5W~0B314qtyGPA2GuZz7EbVI9qIrrL6Wa~a_ejMJjYLL1',
```
Comment (fajarism) [inline next.config.js:13]: CLIENT_ID and CLIENT_SECRET are exposed.
Please delete this.

Comment (fajarism) [inline next.config.js:13]: Since these three are only server side, can we move to serverRuntimeConfig?

File: pages/profile/[username].tsx [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+import LayoutContainer from "../../component/LayoutContainer";
+import Image from "next/image";
+import { Header } from "../../component/Header";
+import Toggle from 'react-toggle'
+import { useGetProfile } from "../../services/profile/profileHooks";
+import { useRouter } from "next/router";
+import { ITokenEnum } from "../../shared/enum";
+import { GetServerSidePropsContext } from "next";
+import { RotatingTriangles } from 'react-loader-spinner'
+import { toast } from "react-toastify";
```
Comment (fajarism) [inline pages/profile/[username].tsx:13]: please change relative import to alias import
e.g 
`import { BannerPage } from "component/Page/BannerPage"`

File: pages/message-sent/index.tsx [hunk 6]
```diff
@@ -25,7 +27,7 @@ export default function MessageSent() {
                         <text className="font-inter font-normal">🕵️‍♀️ Get your own anonymous messages! </text>
                         <text className="font-inter font-normal">🔔 See what’s new in your school or community</text>
                     </div>
-                    <button className=" w-full bg-cyan hover:bg-cyan10 text-white font-lato flex py-2 px-4 rounded-lg justify-center items-center gap-2 text-justify">
+                    <button onClick={()=> window.location.href = publicRuntimeConfig.DOWNLOAD_BETTERSOCIAL_APP_URL} className=" w-full bg-cyan hover:bg-cyan10 text-white font-lato flex py-2 px-4 rounded-lg justify-center items-center gap-2 text-justify">
```
Comment (fajarism) [inline pages/message-sent/index.tsx:30]: any concern why not using next useRouter hook?

File: pages/message-sent/index.tsx [hunk 7]
```diff
@@ -3,7 +3,9 @@ import { Helmet } from "react-helmet";
 import React from "react";
 import LayoutContainer from "../../component/LayoutContainer";
```
Comment (fajarism) [inline pages/message-sent/index.tsx:4]: please change relative import to alias import
e.g
import { BannerPage } from "component/Page/BannerPage"

File: pages/profile/[username].tsx [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+                <div className="flex flex-1-0-0 w-full justify-center items-center">
+                    {/* Card Header */}
+                    {isLoading ? <RotatingTriangles
+                        visible={true}
+                        height="80"
+                        width="80"
+                        ariaLabel="rotating-triangels-loading"
+                        wrapperStyle={{}}
+                        wrapperClass="rotating-triangels-wrapper"
+                    /> : <div className="flex w-full bg-white p-4 rounded-2xl flex-col gap-4 h-max">
```
Comment (fajarism) [inline pages/profile/[username].tsx:45]: Can we make this to a loading wrapper?
e.g.
```
<LoaderWrapper isLoading={isLoading}>
<div className="flex ....>
....
</div>
</LoaderWrapper>
```

So the conditional loading is inside the loader wrapper.

File: pages/verification/index.tsx [hunk 9]
```diff
@@ -3,16 +3,106 @@ import { Helmet } from "react-helmet";
...
 import LayoutContainer from "../../component/LayoutContainer";
 import Image from "next/image";
+import api from "../../shared/fetcher";
+import { useCheckExchangeTokenMutation } from "../../services/auth/authHooks";
+import { useRouter } from "next/router";
+import { GetServerSidePropsContext } from "next";
+import Cookies from "js-cookie";
+import { ITokenEnum } from "../../shared/enum";
+import { useInitChatAnonymousMutation } from "../../services/initChatAnonymous/initChatAnonymousHooks";
+import { toast } from 'react-toastify'
```
Comment (fajarism) [inline pages/verification/index.tsx:13]: please change relative import to alias import
e.g
import { BannerPage } from "component/Page/BannerPage"

File: services/auth/authHooks.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import { UseMutationOptions, useMutation } from "@tanstack/react-query";
+import {
+  CheckExchangeTokenPayload,
+  CheckExchangeTokenResponse,
+  checkExchangeToken,
+} from "./authServices";
+import { AxiosResponse } from "axios";
+import { User } from "../initChatAnonymous/initChatAnonymousServices";
```
Comment (fajarism) [inline services/auth/authHooks.ts:8]: please change relative import to alias import
e.g
import { BannerPage } from "component/Page/BannerPage"

File: services/auth/authServices.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import { AxiosResponse } from "axios";
+import api from "../../shared/fetcher";
+import { User } from "../initChatAnonymous/initChatAnonymousServices";
```
Comment (fajarism) [inline services/auth/authServices.ts:3]: please change relative import to alias import
e.g
import { BannerPage } from "component/Page/BannerPage"

File: shared/fetcher.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
+import axios from "axios";
+import Cookies from "js-cookie";
+
+import getConfig from "next/config";
+import { ITokenEnum } from "./enum";
```
Comment (fajarism) [inline shared/fetcher.ts:5]: please change relative import to alias import
e.g
import { BannerPage } from "component/Page/BannerPage"

File: shared/enum.ts [hunk 13]
```diff
@@ -0,0 +1,9 @@
+export enum ITokenEnum {
+  token = "token",
+  refreshToken = "refresh_token",
+  anonymousToken = "anonymousToken",
+  humanId = "human_id",
+  userId = "user_id",
+  targetUser = "target_user",
+  tempMessage = "temp_message",
```
Comment (fajarism) [inline shared/enum.ts:8]: I think these do not represent TokenEnum.
Please move to another enum
e.g.
`LocalStorageKeyEnum`

**Review by fajarism — CHANGES_REQUESTED** [review]

File: pages/message-sent/index.tsx (hunk 6 above)
Comment (yusufputra) [reply pages/message-sent/index.tsx:30]: actually no, will change to userouter hook

**Comment by yusufputra** [comment]
all updated mas @fajarism

**Review by fajarism — APPROVED** [review]

**Pull request merged**